File: lib/node/alphaskia/src/AlphaSkiaTextStyle.ts
import { AlphaSkiaNative } from './AlphaSkiaNative';
import { AlphaSkiaTypefaceHandle, loadAddon } from './addon';

/**
 * Represents a text style which can be used to draw or measure texts 
 * with support for mixed fonts for fallback character rendering.
 */
export class AlphaSkiaTextStyle extends AlphaSkiaNative<AlphaSkiaTypefaceHandle> {
    /**
     * Gets the list of font family names which are consulted for finding 
     * typefaces with glyphs for drawing or measuring texts.
     */
    public readonly familyNames: string[];

    /**
     * Gets the font weight used for finding typefaces.
     */
    public readonly weight: number;

    /**
     * Gets whether the used typefaces should be italic. 
     */
    public readonly isItalic: boolean;

    /**
     * Initializes a new instance of the {@link AlphaSkiaTextStyle} class.
     * @param familyNames The list of font family names.
     * @param weight The font weight typefaces should have.
     * @param isItalic Whether typefaces should be italic.
     */
    public constructor(familyNames: string[], weight: number, isItalic: boolean) {
        super(
            loadAddon().alphaskia_textstyle_new(AlphaSkiaTextStyle.checkFamilyNames(familyNames), weight, isItalic),
            loadAddon().alphaskia_textstyle_free,
        )
        this.familyNames = familyNames;
        this.weight = weight;
        this.isItalic = isItalic;
    }


    private static checkFamilyNames(familyNames: unknown): string[] {
        if (!Array.isArray(familyNames)) {
            throw new TypeError("family names must be a string array, provided value is no array");
        }

        for (let i = 0; i < familyNames.length; i++) {
            if (typeof familyNames[i] !== 'string') {
                throw new TypeError(`family names must be a string array, provided value at index ${i} is not a string`);
            }
        }

        return familyNames as string[];
    }
}

File: lib/dotnet/AlphaSkia/AlphaSkiaTextStyle.cs
namespace AlphaSkia;

/// <summary>
/// Represents a text style which can be used to draw or measure texts 
/// with support for mixed fonts for fallback character rendering.
/// </summary>
/// <remarks>
/// This structure is designed for good reusability. The same text style might be used 
/// with different font sizes, colors and alignments, that's why they are not part of this structure itself.
/// </remarks>
public sealed class AlphaSkiaTextStyle : AlphaSkiaNative
{
    /// <summary>
    /// Gets the list of font family names which are consulted for finding 
    /// typefaces with glyphs for drawing or measuring texts.
    /// </summary>
    public string[] FontFamilies { get; }

    /// <summary>
    /// Gets the font weight used for finding typefaces.
    /// </summary>
    public ushort Weight { get; }

    /// <summary>
    /// Gets whether the used typefaces should be italic. 
    /// </summary>
    public bool IsItalic { get; }

    /// <summary>
    /// Initializes a new instance of the <see cref="AlphaSkiaTextStyle"/> class.
    /// </summary>
    /// <param name="fontFamilies">The list of font family names.</param>
    /// <param name="weight">The font weight typefaces should have.</param>
    /// <param name="isItalic">Whether typefaces should be italic.</param>
    public AlphaSkiaTextStyle(string[] fontFamilies, ushort weight, bool isItalic)
        : base(NativeMethods.alphaskia_textstyle_new((byte)fontFamilies.Length, fontFamilies, weight, isItalic ? (byte)1 : (byte)0), NativeMethods.alphaskia_textstyle_free)
    {
        FontFamilies = fontFamilies;
        Weight = weight;
        IsItalic = isItalic;
    }
}